kernel/arch/dreamcast/hardware/g2dma.c

```diff
@@ -0,0 +1,234 @@
+        0x00000000: Preserve the current DMA enable setting.
+        0x80000000: Reset the DMA enable setting to "0" after transfer.
+
+    OR '|' this value with length to set the size of DMA request.
+*/
+#define PRESERVE_ENABLED  0x00000000
+#define RESET_ENABLED     0x80000000
+
+/* 
+    Specifies system memory address range for G2-DMA across channels 0-3.
+    Overruns outside this range trigger G2-DMA overrun errors. Previous 
+    setting was 0x4659404f representing range 0x0C400000 - 0x0C4F0000. 
+*/
+#define SYS_MEM_SECURITY_CODE 0x4659
+#define DISABLE_SYS_MEM_PROTECTION (SYS_MEM_SECURITY_CODE << 16 | 0x007F)
+#define ENABLE_SYS_MEM_PROTECTION  (SYS_MEM_SECURITY_CODE << 16 | 0x7F00)
+
+/*  
+    Set the DS# (Data Strobe) timeout to 27 clock cycles for the external DMA. 
+    If data isn't ready for latching by this time, an interrupt will be 
+    triggered. Not sure why its 27 but can be changed later. Default value 
+    is 1023 cycles.
+*/
+#define DS_CYCLE_OVERRIDE  27
+
+/* Disable the DMA */
+static void dma_disable(uint32_t chn) {
+    extdma->dma[chn].enable = 0;
+    extdma->dma[chn].start = 0;
+}
+
+static void g2_dma_irq_hnd(uint32_t code) {
+    int chn = code - ASIC_EVT_G2_DMA0;
+
+    if(chn < 0 || chn > 3) {
+        dbglog(DBG_INFO, "g2_dma: Wrong channel received in g2_dma_irq_hnd");
+        chn = 1;
+    }
+
+    /* VP : changed the order of things so that we can chain dma calls */
+
+    /* Signal the calling thread to continue, if any. */
+    if(dma_blocking[chn]) {
+        sem_signal(&dma_done[chn]);
+        thd_schedule(1, 0);
+        dma_blocking[chn] = 0;
+    }
+
+    /* Call the callback, if any. */
+    if(dma_callback[chn]) {
+        dma_callback[chn](dma_cbdata[chn]);
+    }
+}
+
+int g2_dma_transfer(void * sh4, void * g2bus, uint32_t length, uint32_t block,
+                    g2_dma_callback_t callback, ptr_t cbdata,
+                    uint32_t dir, uint32_t mode, uint32_t g2chn, uint32_t sh4chn) {
+    /* No longer used but we keep then around for compatibility */
+    (void)mode;
+    (void)sh4chn;
+
+    if(g2chn > 3) {
+        errno = EINVAL;
+        return -1;
+    }
+
+    /* Make sure we're not already DMA'ing */
+    if(extdma->dma[g2chn].start != 0) {
+        dbglog(DBG_ERROR, "g2_dma: Already DMA'ing for channel %ld\n", g2chn);
+        errno = EINPROGRESS;
+        return -1;
+    }
+
+    /* Check alignments */
+    if(((uint32_t)sh4) & 31) {
+        dbglog(DBG_ERROR, "g2_dma: Unaligned sh4 DMA %p\n", sh4);
+        errno = EFAULT;
+        return -1;
+    }
+
+    if(((uint32_t)g2bus) & 31) {
+        dbglog(DBG_ERROR, "g2_dma: Unaligned g2bus DMA %p\n", g2bus);
+        errno = EFAULT;
+        return -1;
+    }
+
+    /* Make sure length is a multiple of 32 */
+    length = (length + 0x1f) & ~0x1f;
+
+    dma_blocking[g2chn] = block;
+    dma_callback[g2chn] = callback;
+    dma_cbdata[g2chn] = cbdata;
+
+    /* End current DMA transfer */
+    dma_disable(g2chn);
```

Comment:
Since we check for start=0 before moving to this, would this ever be 'ending a current transfer'?

Reply:
Yeah,  since I added that check to see if the there is a current DMA in progress, this is kinda useless.  Also since I set that extra 31 bit flag on the length, the enable flag automatically gets set to 0 after a transfer ends.  So by the time this function call is reached enable and start should be 0 already.